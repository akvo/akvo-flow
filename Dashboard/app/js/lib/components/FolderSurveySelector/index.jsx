import React from 'react';
import PropTypes from 'prop-types';

export default class FolderSurveySelector extends React.Component {
  state = {
    levels: this.props.levels,
  }

  handleChange = (e) => {
    // pass value back to ember to get new levels
    const { onChange } = this.props;
    const { levels } = this.state;

    const newLevels = onChange(e.target.value);

    if (newLevels) {
      this.setState({ levels: levels.concat(newLevels) });
    }
  }

  renderForm = (folderSurveyList, id) => (
    <select data-testid={`folder-survey-select-${id}`} key={id} onChange={this.handleChange}>
      {folderSurveyList.map(surveyGroup => (
        <option key={surveyGroup.keyId} value={surveyGroup.keyId}>
          {surveyGroup.name}
        </option>
      ))}
    </select>
  )

  render() {
    const { levels } = this.state;

    return (
      <div>
        {/* foreach levels in array render a new select form */}
        {levels.map(this.renderForm)}
      </div>
    );
  }
}

FolderSurveySelector.propTypes = {
  levels: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
};
