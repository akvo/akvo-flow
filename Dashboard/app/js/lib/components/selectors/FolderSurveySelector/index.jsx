import React from 'react';
import PropTypes from 'prop-types';

export default class FolderSurveySelector extends React.Component {
  state = {
    levels: [],
  };

  componentDidMount() {
    const { initialSurveyGroup, surveyGroups } = this.props;

    if (initialSurveyGroup) {
      // if initial form is available, generate levels and mark it as selected
      const { parentId } = surveyGroups.find(sg => sg.keyId == initialSurveyGroup);

      this.setState({ levels: this.getLevels(parentId) }, () => {
        this.props.onSelectSurvey(initialSurveyGroup);
      });
    } else {
      this.setState({ levels: this.getLevels(0) });
    }
  }

  getLevels = (parentId = 0) => {
    const { surveyGroups, strings, initialSurveyGroup } = this.props;
    const initialSurvey = surveyGroups.find(sg => sg.keyId == initialSurveyGroup);
    const levels = [];

    // eslint-disable-next-line eqeqeq
    if (parentId != 0) {
      // eslint-disable-next-line eqeqeq
      const parent = surveyGroups.find(sg => sg.keyId == parentId);

      for (let i = 0; i < parent.ancestorIds.length; i++) {
        const parentSurveys = surveyGroups
          .reduce((total, sgs) => {
            // eslint-disable-next-line eqeqeq
            if (sgs.parentId == parent.ancestorIds[i]) {
              return total.concat({
                ...sgs,
                selected: initialSurvey && initialSurvey.ancestorIds.includes(sgs.keyId),
              });
            }

            return total;
          }, [])
          .sort(this.comparator);

        const level = [
          {
            keyId: 0,
            parentId: null,
            name: strings.chooseFolderOrSurvey,
          },
        ].concat(parentSurveys);

        levels.push(level);
      }
    }

    const surveys = surveyGroups
      .reduce((total, sgs) => {
        // eslint-disable-next-line eqeqeq
        if (sgs.parentId == parentId) {
          return total.concat({
            ...sgs,
            selected: initialSurvey && initialSurvey.keyId == sgs.keyId,
          });
        }

        return total;
      }, [])
      .sort(this.comparator);

    levels.push(
      [
        {
          keyId: 0,
          parentId: null,
          name: strings.chooseFolderOrSurvey,
        },
      ]
        // eslint-disable-next-line eqeqeq
        .concat(surveys)
    );

    return levels;
  };

  comparator = (a, b) => {
    const nameA = (a.name || '').toUpperCase(); // ignore upper and lowercase
    const nameB = (b.name || '').toUpperCase(); // ignore upper and lowercase

    if (nameA < nameB) {
      return -1;
    }
    if (nameA > nameB) {
      return 1;
    }

    // names must be equal
    return 0;
  };

  handleChange = e => {
    const parentId = e.target.value;

    // check if a survey has been selected
    // eslint-disable-next-line eqeqeq
    if (!this.props.onSelectSurvey(parentId)) {
      return null;
    }

    const newLevels = this.getLevels(parentId);

    if (newLevels) {
      this.setState({ levels: newLevels });
    }

    return null;
  };

  renderForm = (folderSurveyList, id) => {
    const defaultSurveyGroup = folderSurveyList.find(sgs => sgs.selected === true);

    return (
      <select
        defaultValue={defaultSurveyGroup ? defaultSurveyGroup.keyId : undefined}
        data-testid={`folder-survey-select-${id}`}
        key={id}
        onChange={this.handleChange}
      >
        {folderSurveyList.map(surveyGroup => (
          <option key={surveyGroup.keyId} value={surveyGroup.keyId}>
            {surveyGroup.name}
          </option>
        ))}
      </select>
    );
  };

  render() {
    const { levels } = this.state;

    return (
      <div>
        {/* foreach levels in array render a new select form */}
        {levels.map(this.renderForm)}
      </div>
    );
  }
}

FolderSurveySelector.propTypes = {
  surveyGroups: PropTypes.array.isRequired,
  onSelectSurvey: PropTypes.func.isRequired,
  strings: PropTypes.object.isRequired,
  initialSurveyGroup: PropTypes.any, // eslint-disable-line react/require-default-props
};
