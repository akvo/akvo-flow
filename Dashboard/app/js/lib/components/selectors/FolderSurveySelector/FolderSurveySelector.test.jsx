import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import FolderSurveySelector from './index';
import '@testing-library/jest-dom/extend-expect';

// https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
const originalError = console.error;
beforeAll(() => {
  console.error = (...args) => {
    if (/supports * the "act"/.test(args[0])) {
      return;
    }
    originalError.call(console, ...args);
  };
});

afterAll(() => {
  console.error = originalError;
});


// Main Tests
describe('FolderSurveySelector Tests', () => {
  afterEach(cleanup);


  const initialLevels = [[
    { keyId: 0, name: 'Choose a folder or survey', parentId: null },
    { keyId: 153142013, name: 'Folder with few large data sets', parentId: 0 },
  ]];

  it('+++ renders <snapshot>', () => {
    const wrapper = render(<FolderSurveySelector onChange={jest.fn} levels={initialLevels} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('+++ calls for new levels and update state when option is selected', () => {
    // create mock function and make it return new levels when called
    const onChangeFn = jest.fn();
    onChangeFn.mockReturnValue([[
      { keyId: 0, parentId: null, name: 'Choose a folder or survey' },
      { keyId: 152342023, parentId: 153142013, name: 'BAR-handpump' },
    ]]);

    const wrapper = render(<FolderSurveySelector onChange={onChangeFn} levels={initialLevels} />);

    // select a select form and trigger a value change
    const select = wrapper.getByTestId('folder-survey-select-0');
    fireEvent.change(select, { target: { value: 153142013 } });

    // expect the onChange props to have been called once
    expect(onChangeFn).toHaveBeenCalledTimes(1);

    // expect the onChange props to be called with the right value
    expect(onChangeFn).toHaveBeenCalledWith('153142013');

    // expect a new select form to be available
    expect(wrapper.getByTestId('folder-survey-select-1')).toBeTruthy();

    // save snapshot of the current result
    expect(wrapper.container).toMatchSnapshot();
  });

  it('+++ doesn\'t update state when no new value comes in after option is selected', () => {
    // create mock function and make it return null when called
    const onChangeFn = jest.fn();
    onChangeFn.mockReturnValue(null);

    const wrapper = render(<FolderSurveySelector onChange={onChangeFn} levels={initialLevels} />);

    // select a select form and trigger a value change
    const select = wrapper.getByTestId('folder-survey-select-0');
    fireEvent.change(select, { target: { value: 153142013 } });

    // expect the onChange props to have been called once
    expect(onChangeFn).toHaveBeenCalledTimes(1);

    // expect the onChange props to be called with the right value
    expect(onChangeFn).toHaveBeenCalledWith('153142013');

    // expect no new select form to be available
    expect(wrapper.queryByTestId('folder-survey-select-1')).toBeFalsy();
  });
});
