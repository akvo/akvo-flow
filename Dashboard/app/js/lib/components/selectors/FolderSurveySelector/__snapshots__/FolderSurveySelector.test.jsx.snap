// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderSurveySelector Tests +++ checks if a survey is selected when options is clicked 1`] = `
<div>
  <div>
    <select
      data-testid="folder-survey-select-0"
    >
      <option
        value="0"
      >
        Choose folder or survey
      </option>
      <option
        value="153142013"
      >
        Folder with a few large data sets
      </option>
    </select>
    <select
      data-testid="folder-survey-select-1"
    >
      <option
        value="0"
      >
        Choose folder or survey
      </option>
      <option
        value="152342023"
      >
        BAR-handpump
      </option>
      <option
        value="148412306"
      >
        NR-handpump
      </option>
    </select>
  </div>
</div>
`;

exports[`FolderSurveySelector Tests +++ renders <snapshot> 1`] = `
<div>
  <div>
    <select
      data-testid="folder-survey-select-0"
    >
      <option
        value="0"
      >
        Choose folder or survey
      </option>
      <option
        value="153142013"
      >
        Folder with a few large data sets
      </option>
    </select>
  </div>
</div>
`;
